File: app/src/main/java/closer/vlllage/com/closer/PhotoActivity.java
package closer.vlllage.com.closer;

import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.WindowManager;

import com.github.chrisbanes.photoview.PhotoView;
import com.squareup.picasso.Callback;

import closer.vlllage.com.closer.handler.helpers.DefaultAlerts;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.ImageHandler;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.helpers.SystemShareHandler;
import closer.vlllage.com.closer.ui.CircularRevealActivity;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.BehaviorSubject;
import jp.wasabeef.picasso.transformations.RoundedCornersTransformation;

public class PhotoActivity extends CircularRevealActivity {
    public static final String EXTRA_PHOTO = "photo";
    private PhotoView photo;
    private BehaviorSubject<Boolean> enterAnimationCompleteObservable = BehaviorSubject.create();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_photo);
        photo = findViewById(R.id.photo);
        photo.setMaximumScale(8f);

        enterAnimationCompleteObservable.onNext(true);

        if (getIntent() != null) {
            String photoUrl = getIntent().getStringExtra(EXTRA_PHOTO);
            $(ImageHandler.class).get().load(photoUrl)
                    .transform(new RoundedCornersTransformation($(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.imageCorners), 0))
                    .into(photo, new Callback() {
                        @Override
                        public void onSuccess() {
                            $(DisposableHandler.class).add(enterAnimationCompleteObservable
                                    .subscribeOn(Schedulers.computation())
                                    .observeOn(AndroidSchedulers.mainThread())
                                    .subscribe(value -> {
                                        loadFullRes(photoUrl);
                                    }, Throwable::printStackTrace));
                        }

                        @Override
                        public void onError(Exception e) {
                            onSuccess();
                        }
                    });
        }

        photo.setOnClickListener(view -> finish());

        findViewById(R.id.actionShare).setOnClickListener(view -> {
            if (photo.getDrawable() instanceof BitmapDrawable) {
                $(SystemShareHandler.class).share(((BitmapDrawable) photo.getDrawable()).getBitmap());
            } else {
                $(DefaultAlerts.class).thatDidntWork();
            }
        });
    }

    @Override
    protected int getBackgroundId() {
        return R.id.activityLayout;
    }

    private void loadFullRes(String photoUrl) {
        $(ImageHandler.class).get().load(photoUrl.split("\\?")[0] + "?s=1600")
                .noPlaceholder()
                .transform(new RoundedCornersTransformation($(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.imageCorners), 0))
                .into(photo);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/PhotoActivity.kt
package closer.vlllage.com.closer

import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.ui.CircularRevealActivity
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.github.chrisbanes.photoview.PhotoView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import kotlin.math.abs

class PhotoActivity : CircularRevealActivity() {

    private lateinit var photo: PhotoView
    private val enterAnimationCompleteObservable = BehaviorSubject.create<Boolean>()

    override fun onCreate(savedInstanceState: Bundle?) {
        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN or WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS, WindowManager.LayoutParams.FLAG_FULLSCREEN)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            val attrib = window.attributes
            attrib.layoutInDisplayCutoutMode = WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES
        }

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_photo)
        photo = findViewById(R.id.photo)
        photo.maximumScale = 8f

        photo.setOnSingleFlingListener { _, _, velocityX, velocityY ->
            if (photo.scale != 1f) {
                return@setOnSingleFlingListener false
            }

            if (abs(velocityX) > SLOP_RADIUS || abs(velocityY) > SLOP_RADIUS) {
                finish()
                return@setOnSingleFlingListener true
            }

            return@setOnSingleFlingListener false
        }

        enterAnimationCompleteObservable.onNext(true)

        if (intent != null) {
            val photoUrl = intent.getStringExtra(EXTRA_PHOTO)!!
            on<ImageHandler>().get().load(photoUrl).listener(object : RequestListener<Drawable> {
                override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<Drawable>?, isFirstResource: Boolean): Boolean {
                    onSuccess(photoUrl)
                    return false
                }

                override fun onResourceReady(resource: Drawable?, model: Any?, target: Target<Drawable>?, dataSource: DataSource?, isFirstResource: Boolean): Boolean {
                    onSuccess(photoUrl)
                    return false
                }

            }).into(photo)
        }

        photo.setOnClickListener { finish() }

        findViewById<View>(R.id.actionShare).setOnClickListener {
            if (photo.drawable is BitmapDrawable) {
                on<SystemShareHandler>().share((photo.drawable as BitmapDrawable).bitmap)
            } else {
                on<DefaultAlerts>().thatDidntWork()
            }
        }
    }

    override val backgroundId = R.id.activityLayout

    private fun onSuccess(photoUrl: String) {
        on<DisposableHandler>().add(enterAnimationCompleteObservable
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ loadFullRes(photoUrl) }, {
                    on<ToastHandler>().show(R.string.failed_to_load_photo)
                }))
    }

    private fun loadFullRes(photoUrl: String) {
        on<ImageHandler>().get().load(photoUrl.split("\\?".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0] + "?s=1600")
                .apply(RequestOptions().skipMemoryCache(true))
                .placeholder(photo.drawable)
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(photo)
    }

    companion object {
        const val EXTRA_PHOTO = "photo"
        private const val SLOP_RADIUS = 32
    }
}
